mountpoint-s3/src/checksums/block.rs

```diff
@@ -0,0 +1,213 @@
+use bytes::{Bytes, BytesMut};
+use mountpoint_s3_crt::checksums::crc32c::{self, Crc32c};
+
+use crate::checksums::{bytes::ChecksummedBytes, combine_checksums, IntegrityError};
+
+/// A `ChecksummedBlock` is a bytes buffer that carries its checksum.
+/// The implementation guarantees that its integrity will be validated when data is accessed.
+#[derive(Debug, Clone)]
+pub struct ChecksummedBlock {
```

Comment:
I was wondering if it might be nicer to have just one implementation of this stuff, and give `ChecksummedBytes` a [`shrink_to_fit`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.shrink_to_fit)-style method to get the guarantee you're looking for. But then I guess that makes `extend` et al more complicated because you have to handle all the different combinations to decide when you can skip validating the checksums, so probably not worth it?

Reply:
After all, I think `shrink_to_fit` would be a better approach and can also be used to improve `extend`. I will close this PR and open a new one with that change.